### ISSUE TITLE
CorpseFinder Uninstall-Watcher notifications use out-dated (previous?) app infos 

### ISSUE THREAD

**Issue by d4rken** [d4rken-org/sdmaid-se#585]
One last post for today: Metadata Remover was deinstalled days ago. I've deinstalled five more apps after it, documented by the app "Installalogs". How comes this string is still being used in the notification?

v0.10.8-beta1
Android 10
Fairphone FP3
Play Services updated to v23.13.12

_Originally posted by @uli-on in https://github.com/d4rken-org/sdmaid-se/issues/569#issuecomment-1666972218_

**Comment by d4rken** [comment]
@uli-on 

> SD Maid opens a plethora of notifications while executing the Corpse Finder, at least three. Are you sure you clicked through to the last one even if you saw "0 objects to remove" on one of the earlier notifications?

I'm 95% sure that I was checking with the deletion notification. Because I was also trying to see if I can display the app name, only to notice that the app name is obviously unavailable, as the uninstall watcher triggers on "app was uninstalled" events, at which point it is already too late to get the app name.

Can you provide a video, or record a debug log of the whole thing, or maybe a step by step instruction, like: Install app A, install app B, open app A and B, remove app A, then app B, and notice that the notification shows app A's package name not app B's? That's what I tried and it worked for me. I tested on Android 12 though, maybe this is API specific? :thinking:

**Comment by uli-on** [comment]
Log file is in the wire.

**Comment by uli-on** [comment]
Forgot to mention in my email that I started SD Maid once after updating it to v0.11.0—beta0 and swiped it out of the app switcher so it might start afresh after the update, and also swiped the "Success!" notification out of the notification drawer to give it the chance to refresh the information displayed.

**Comment by d4rken** [comment]
Thanks for the most recent log Uli. Going through it at the moment.

SD Maid gets 3 events from the system, which should mean 3 uninstalls:

```
1st one: org.billthefarmer.buses
2023-08-10T13:13:02.879Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Receiver: onReceive(android.app.ReceiverRestrictedContext@dc06b89,Intent { act=android.intent.action.PACKAGE_FULLY_REMOVED dat=package:org.billthefarmer.buses flg=0x5000010 cmp=eu.darken.sdmse/.corpsefinder.core.watcher.UninstallWatcherReceiver (has extras) })

2nd one: org.billthefarmer.solver 
2023-08-10T13:14:19.584Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Receiver: onReceive(android.app.ReceiverRestrictedContext@dc06b89,Intent { act=android.intent.action.PACKAGE_FULLY_REMOVED dat=package:org.billthefarmer.solver flg=0x5000010 cmp=eu.darken.sdmse/.corpsefinder.core.watcher.UninstallWatcherReceiver (has extras) })

3rd one: org.billthefarmer.currency 
2023-08-10T13:15:32.917Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Receiver: onReceive(android.app.ReceiverRestrictedContext@dc06b89,Intent { act=android.intent.action.PACKAGE_FULLY_REMOVED dat=package:org.billthefarmer.currency flg=0x5000010 cmp=eu.darken.sdmse/.corpsefinder.core.watcher.UninstallWatcherReceiver (has extras) })
```

3 different apps too, so far so good.

Let's check the notifications:

```
2023-08-10T13:13:16.013Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfScan(Scan(pkgId=org.billthefarmer.buses, foundItems=0))
2023-08-10T13:13:39.519Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfDeletion(Deletion(appName=null, pkgId=com.demizo.daily_you, deletedItems=0, freedSpace=0))
2023-08-10T13:14:27.819Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfScan(Scan(pkgId=org.billthefarmer.solver, foundItems=0))
2023-08-10T13:14:43.366Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfDeletion(Deletion(appName=null, pkgId=com.demizo.daily_you, deletedItems=0, freedSpace=0))
2023-08-10T13:15:40.344Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfScan(Scan(pkgId=org.billthefarmer.currency, foundItems=0))
2023-08-10T13:16:05.179Z  D/SDMSE:CorpseFinder:Watcher:Uninstall:Notifications: notifyOfDeletion(Deletion(appName=null, pkgId=com.demizo.daily_you, deletedItems=0, freedSpace=0))
```

Suddenly we have notifications for `com.demizo.daily_you`. Where does this come from... hm

First mention in the logs is an SD Maid internal forwarded intent:

```
2023-08-10T13:13:31.818Z  D/SDMSE:CorpseFinder:Watcher:Task:Receiver: onReceive(android.app.ReceiverRestrictedContext@dc06b89,Intent { act=corpsefinder.watcher.uninstall.intent.NEW_TASK flg=0x10000010 cmp=eu.darken.sdmse/.corpsefinder.core.watcher.ExternalWatcherTaskReceiver bnds=[24,1493][438,1637] (has extras) })
2023-08-10T13:13:31.819Z  I/SDMSE:CorpseFinder:Watcher:Task:Receiver: Received task is Delete(target=com.demizo.daily_you)
```

Hm so we have auto deletion enabled for the watcher, and for every scan of the legit apps, the delete task is not for the app that was just deleted but for some reason for `com.demizo.daily_you`.

So the delete task that comes after the scan has the wrong arguments.
The notification that shows the results of the "scan task" and has a "delete remnants" action does not update it's arguments, but instead sends the same one from the ? last reboot ?. The deletion target (the pkg name) is what is displayed in the deletion result notification and thus is wrong.

So we need to make sure the notification is updated with right arguments :beers:

**Comment by d4rken** [comment]
Thanks for your help. The fix in #598 will be included in the next update.

**Comment by uli-on** [comment]
Is the fix for this in v0.11.0 beta0? Then it's not squished completely. I still have an ancient packagename displayed. Might be due to the last notification ("cleanup successful, nB released, n items successfully removed") not being built anew when called: I see a (German) "2T" (=2 days) in the notification's header.

**Comment by d4rken** [comment]
No, only in next update, which was delayed so far due to the Google stuff. I hope to release it soon.
